src/Guards/TokenGuard.php

```diff
@@ -82,6 +82,24 @@ public function __construct(ResourceServer $server,
         $this->encrypter = $encrypter;
     }
 
+    /**
+     * Determine if the requested provider matches the client's provider.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @return bool
+     */
+    protected function validateProvider(Request $request)
+    {
+        $client = $this->client($request);
+
+        // If not client provider is defined, fallback to old behavior.
+        if ($client && empty($client->getProvider())) {
+            return true;
+        }
+
+        return $client && $this->provider == $client->getProvider();
```

Comment:
I still think we should a different comparison here instead of comparing two objects but I'm not sure what exact. Can we compare string names from the providers?

Reply:
I would like the do a more strict comparison here as well but I don't really have a good idea on how to get the name of the provider. The TokenGuard receives the requested UserProvider already and there isn't a string to pull from on the UserProvider class to my knowledge.

Reply:
One solution I guess would be to rewrite the TokenGuard constructor to accept the string instead of the actual UserProvider then convert it to a UserProvider inside the TokenGuard based on the oauth_client provider. This would touch more code/tests though.

Reply:
Hmm, I think it's a simple as this?

```suggestion
        return $client && $this->provider === $client->provider;
```

Reply:
I think this is what we discussed before, using `===` wouldn't work here because we would be comparing instances of the class. We could have different instances of the same class here so it would fail.

`$this->provider` is a `UserRepository` passed into the `TokenGuard` from the PassportServiceProvider based on the `RequestGuard`.
`$client->provider` is a string defined on the `oauth_client`.
`$client->getProvider()` could be a `UserRepository` but it would be its own instance.

Reply:
I see, thanks for clarifying.

I'm actually uncertain why this check is needed? Wouldn't this request prevent any authentication on clients that have a different provider than the default one set through the config? Maybe I'm not entirely following what's going on.

Reply:
So let me try to clarify.

When the PassportServiceContainer loads the TokenGuard it is passing in the UserProvider from the RequestGuard. This is sort of the crux of the problem as this can allow a token created by a Customer to receive a User instance.

What I'm doing is allowing that UserProvider to still be passed in but instead of just trusting it and returning a user from it, I check if the oauth_client has a provider defined. If a provider is defined on the oauth_client then we use that one and compare it (loosely) to the one passed in from the constructor. If there is a mismatch then no user is returned and thus you are unauthenticated. If no provider is defined on the oauth_client then everything continues to work as it always has.

Reply:
But am I right that `$this->provider` only holds **the default provider set from the config**? This means that no client with a different custom provider will get through the check? Maybe I'm still misunderstanding something.

Reply:
I did not change the way the default provider is passed into the TokenGuard. The UserProvider getting passed in has no context of the Laravel\Passport\Client. So it is passing in whatever provider the RequestGuard is using. Since this can be a mismatch to the one now defined on the Laravel\Passport\Client we need to check it inside the TokenGuard.

I can setup a demo to better showcase the problem if you want. I'm also available to discuss on discord if that works for you.

Reply:
@billriess
I think there is another way to validate the provider by comparing the table of client provider against the table of user entity retrieved using oauth_user_id

$providers = config('auth.providers');
$clientProvider = $providers[$client->provider];
$table = $clientProvider['table'] ??
         (new $clientProvider['model'])->getTable();

And compare this table against that of the already retrieved user.

Reply:
I think we're talking about different things...

You can look at the demo here:
https://github.com/billriess/passport-test

There is a sqlite db that you need to point to in your .env file.

Then you can make a REST call to /oauth/token using:
```js
{
    "grant_type": "password",
    "client_id": "1",
    "client_secret": "secret",
    "username": "user@test.test",
    "password": "password"
}
```
and
```js
{
    "grant_type": "password",
    "client_id": "2",
    "client_secret": "secret",
    "username": "customer@test.test",
    "password": "password"
}
```

Now use those tokens and hit `/api/user` or `/api/customer`. You'll see that `auth()->user()` will always return a `User` for the first endpoint and a `Customer` for the second endpoint even if you're using a token from the other model. The token represents who is authenticated so you shouldn't get an instance of `User` if you're token originated from `Customer`.

Reply:
I know what you are talking about because I've addressed the same problem in my package.
See the auhtenticateViaBearerToken here:
https://github.com/hamedov93/passport-multiauth/blob/master/src/Guards/TokenGuard.php

Reply:
Yeah, I guess it just seems odd to me to retrieve the user at all if it's not the right user.

Reply:
The user will be retrieved any way for successful requests.

The cases when the provider doesn't match should be rare or suspecious, so another database query wouldn't be a big deal.

Reply:
Yeah, I guess there are a bunch of different ways to tackle the problem. I think what I have is fairly clean though. We're still comparing the classes just not the instances.

Reply:
I think we should refactor the way the user is retrieved in general:

- Check if the client from the token has a provider
- If it has a provider, use that one to retrieve the user
- If not, use the default one which is passed in the token guard

That way you don't need to do any comparison.

Reply:
@billriess feel free to ping me on Discord in the passport channel

Reply:
Ok, I'll look for you on discord.

I think always returning the user can be an issue though. If the two mismatch you want to not return any user because that means the token they are using is not the same as the one being protected by the route's guard.

Reply:
@driesvints I think it's important to understand why different providers should not return any user. If you protected a route using the `auth:api:customers` middleware it would always return the provider for the customers. This, by default, would ensure that the user authenticated would be a customer. But, there is no check in place to make sure that the auth token actually belongs to the customer provider. So you end up in a scenario where a user auth token would return a customer instead of the expected user.

What I do is compare the 2 different providers. The provider on the RequestGuard is saying this must be a customer. Then inside the Passport TokenGuard I look at the the auth token's client and see if there is a provider defined. If there is, then we check if it matches the same class as the one from the RequestGuard. If they match, we can return the user because we know the user_id would come from the right provider. If they mismatch, then we have an authentication problem because you're trying to access a customer when you're token is not set up for that.

Reply:
I now fully understand what you mean. Thanks for your patience and explaining to well :-)

I think we should indeed pass the provider name to the token guard as well and do the comparison like that. That will be the only way to fully validate it. Comparing objects like this will lead to bugs I believe.

Reply:
We could add a `getName` method to the `UserProvider` contract but that would need a new major version of the framework. Passing the provider name to the TokenGuard is a good workaround.

Reply:
Ok but this will end up with a fairly large amount of test rewrites because they are all expecting a provider to be passed so they currently anticipate the user.

A `getName` method would be great but yes that would require a framework modification to 8.x. I can push a PR for that as well if desired.

Reply:
@billriess I'll run this over with Taylor

Reply:
Any update here? I don't mind making changes to the PR but it would be helpful to know what direction you or Taylor would like to take this in. This is one of the most requested/discussed features of Passport and as I see it, could be a potential security concern for some inexperienced developers.

@driesvints @taylorotwell

Reply:
Actually this doesn't require a major release to the framework, as the provider can be reversely figured out using the model in EloquentUserProvider or the table in DatabaseUserProvider.
You already have a model instance or the table name, so you can fetch the provider from config that uses the model or the table name.

Reply:
What if it uses something other than those two? I don't think it would be safe to assume those are the only two types of provider implementations.

Reply:
There is a method that creates the provider class for each driver.
`createEloquentDriver` and `createDatabaseDriver`
If another driver is added there would be another method to be implemented and therefore passport will be updated.
I am not suggesting to use this approach forever, but to avoid the hassle of new major version for just a trivial need. But it should be added to the next release any way.

Reply:
Looks like passport is using eloquent driver explicitly in the code so the is no need to check for database driver.
An easy solution for this would be like this:
```
$model = config('auth.providers.'.$client->provider.'.model');
if (class_exists($model))
{
    // Before calling retrieveById();
    $this->provider->setModel($model);
}
else
{
  // Other wise request should fail
}
```
I think this is enough :)

Reply:
```php
/**
     * Determine if the requested provider matches the client's provider.
     *
     * @param  \Illuminate\Http\Request  $request
     * @return bool
     */
    protected function hasValidProvider(Request $request)
    {
        $client = $this->client($request);

        // If no client provider is defined, fallback to old behavior.
        if ($client && empty($client->getProvider())) {
            return true;
        }

        // Determine if the client's provider and the request's provider have matching models.
        return $client && $client->getProvider()->getModel() === $this->provider->getModel();
    }
```

How's this? Same outcome with strict comparison.

Reply:
@billriess can you give me push access to the PR? (see Discord)